### PR TITLE
add with_musl compile option to support compile in musl-linux such as alpine

### PR DESCRIPTION
**Pull request by chen-zhiyu**
### PR types

New features

### PR changes

Others

### Describe

to support port paddle to  libos-like SGX TEE enironment, which used musl as basic libc runtime.
currently supported scone and occlum, and graphene is under working.

1. add with_musl compile option
2. fix problem in quering libc version
3. add macro to disable involved the backtrace(which is not valid in musl) when with_musl=on
4. upgrade protobuf from 3.1.0 to 3.5.0, to fix the bswap_* function redefined in probuff lib when with_crypto is enabled.

### CODE REVIEW

File: cmake/configure.cmake [hunk 1]
```diff
@@ -51,6 +51,16 @@ if(WIN32)
   endif(NOT MSVC)
 endif(WIN32)
 
+if(WITH_MUSL)
+    add_definitions(-DPADDLE_WITH_MUSL)
+
+    message(STATUS, "WITH_MKL=OFF, when WITH_MUSL=ON")
```
Comment (chenwhql) [inline cmake/configure.cmake]: recommended to write more like an English message, such as `Set conpile option WITH_MKL=OFF when compile option WITH_MUSL=ON`

File: cmake/configure.cmake [hunk 2]
```diff
@@ -51,6 +51,16 @@ if(WIN32)
...
+    SET(WITH_MKL OFF)
+
+    message(STATUS, "WITH_GPU=OFF, when WITH_MUSL=ON")
```
Comment (chenwhql) [inline cmake/configure.cmake]: same above

File: paddle/fluid/platform/enforce.h [hunk 3]
```diff
@@ -230,13 +230,13 @@ inline std::string SimplifyDemangleStr(std::string str) {
 }
 
 inline std::string GetCurrentTraceBackString() {
-  static constexpr int TRACE_STACK_LIMIT = 100;
+  //static constexpr int TRACE_STACK_LIMIT = 100;
```
Comment (chenwhql) [inline paddle/fluid/platform/enforce.h]: I think here we can use `#ifdef PADDLE_WITH_MUSL` control the backtrace related code, now paddle still need backtrace in many cases.

File: paddle/fluid/platform/macros.h [hunk 4]
```diff
@@ -25,6 +25,6 @@ limitations under the License. */
   classname& operator=(classname&&) = delete
 #endif
 
-#if defined(__FLT_MAX__)
-#define FLT_MAX __FLT_MAX__
-#endif  // __FLT_MAX__
+//#if defined(__FLT_MAX__)
```
Comment (chenwhql) [inline paddle/fluid/platform/macros.h]: same above, also need to use `#ifdef PADDLE_WITH_MUSL`, avoid introducing other problems

File: paddle/fluid/platform/port.h [hunk 5]
```diff
@@ -26,7 +26,7 @@
 
 #if !defined(_WIN32)
 #include <dlfcn.h>     //  dladdr
-#include <execinfo.h>  // backtrace
+//#include <execinfo.h>  // backtrace
```
Comment (chenwhql) [inline paddle/fluid/platform/port.h]: use `#ifndef PADDLE_WITH_MUSL`

File: cmake/external/protobuf.cmake [hunk 6]
```diff
@@ -234,7 +234,7 @@ FUNCTION(build_protobuf TARGET_NAME BUILD_FOR_HOST)
     )
 ENDFUNCTION()
 
-SET(PROTOBUF_VERSION 3.1.0)
+SET(PROTOBUF_VERSION 3.5.0)
```
Comment (chenwhql) [inline cmake/external/protobuf.cmake]: need @luotao1 to evaluate the version upgrade here

File: paddle/fluid/platform/enforce.h (hunk 3 above)
Comment (chen-zhiyu) [reply paddle/fluid/platform/enforce.h]: ok

File: paddle/fluid/platform/port.h (hunk 5 above)
Comment (chen-zhiyu) [reply paddle/fluid/platform/port.h]: ready in 2e12aebf246b69294ff57ee797105b43ab288b0e

File: cmake/external/protobuf.cmake (hunk 6 above)
Comment (chen-zhiyu) [reply cmake/external/protobuf.cmake:237]: ready in 2e12aebf246b69294ff57ee797105b43ab288b0e

File: paddle/fluid/platform/macros.h (hunk 4 above)
Comment (chen-zhiyu) [reply paddle/fluid/platform/macros.h]: ready in 2e12aebf246b69294ff57ee797105b43ab288b0e

File: paddle/fluid/platform/enforce.h (hunk 3 above)
Comment (chen-zhiyu) [reply paddle/fluid/platform/enforce.h]: ready in 2e12aebf246b69294ff57ee797105b43ab288b0e

File: cmake/configure.cmake (hunk 2 above)
Comment (chen-zhiyu) [reply cmake/configure.cmake]: ready in 2e12aebf246b69294ff57ee797105b43ab288b0e

File: cmake/configure.cmake (hunk 1 above)
Comment (chen-zhiyu) [reply cmake/configure.cmake]: ready in 2e12aebf246b69294ff57ee797105b43ab288b0e

File: paddle/scripts/musl_build/Dockerfile [hunk 7]
```diff
@@ -0,0 +1,12 @@
+FROM python:3.7-alpine3.10
+
+RUN apk add --no-cache --force-overwrite\
+    g++  gfortran make cmake patchelf\
+    git \
+    linux-headers \
+    libjpeg-turbo-dev \ 
+    lapack-dev openblas-dev 
+    
+#RUN pip3 install scikit-build\
```
Comment (chenwhql) [inline paddle/scripts/musl_build/Dockerfile]: 这个注释不需要的话是不是先删掉？以及文件最后一般要回车到空白行

Comment (chen-zhiyu) [reply paddle/scripts/musl_build/Dockerfile]: 我先去掉了这个文件，下次和docker编译脚本一起提

**Review by chenwhql — APPROVED** [review]
LGTM

**Review by luotao1 — APPROVED** [review]
LGTM

**Comment by YueWang1996** [comment]
您好，我按照您的https://github.com/chen-zhiyu/Paddle/tree/develop/paddle/scripts/musl_build “Build Manually”教程成功安装了paddle, 然后我尝试在自己的alpine-python3.7镜像上安装paddle失败了，我在cmake时遇到以下问题并没有找到解决办法，如果可以拜托您可以帮我看一下，谢谢~
```
~/build # cmake -DWITH_MUSL=ON 
-DWITH_CRYPTO=OFF 
-DWITH_MKL=OFF -DWITH_GPU=OFF 
-DPYTHON_INCLUDE_DIR=/usr/include/python3.7m 
-DPYTHON_LIBRARY=/usr/lib/libpython3.so 
-DPY_PIP=/usr/lib/python3.7/site-packages/pip 
-DPY_NUMPY=/usr/lib/python3.7/site-packages/numpy 
-DPY_WHEEL=/usr/lib/python3.7/site-packages/wheel 
-DPY_GOOGLE.PROTOBUF=/usr/lib/python3.7/site-packages/google/protobuf 
-DPYTHON_NUMPY_INCLUDE_DIR=/usr/lib/python3.7/site-packages/numpy/core/include/numpy /paddle
-- Found Paddle host system: , version: 3.10
-- Found Paddle host system's CPU: 8 cores
-- CXX compiler: /usr/bin/c++, version: GNU 8.3.0
-- C compiler: /usr/bin/cc, version: GNU 8.3.0
-- AR tools: /usr/bin/ar
-- Do not have AVX2 intrinsics and disabled MKL-DNN
-- BOOST_VERSION: 1.41.0, BOOST_URL: http://paddlepaddledeps.bj.bcebos.com/boost_1_41_0.tar.gz
-- warp-ctc library: /root/build/third_party/install/warpctc/lib/libwarpctc.so
-- Found OpenBLAS (include: /usr/include, library: /usr/lib/libopenblas.so)
-- Found lapack in OpenBLAS (include: /usr/include)
-- Protobuf protoc executable: /root/build/third_party/install/protobuf/bin/protoc
-- Protobuf-lite library: /root/build/third_party/install/protobuf/lib/libprotobuf-lite.a
-- Protobuf library: /root/build/third_party/install/protobuf/lib/libprotobuf.a
-- Protoc library: /root/build/third_party/install/protobuf/lib/libprotoc.a
-- Protobuf version: 3.1.0
CMake Error at cmake/external/python.cmake:74 (MESSAGE):
  Found Python Protobuf < 3.0.0, please use pip to upgrade protobuf.  pip
  install -U protobuf
Call Stack (most recent call first):
  cmake/third_party.cmake:236 (include)
  CMakeLists.txt:228 (include)


-- Configuring incomplete, errors occurred!
See also "/root/build/CMakeFiles/CMakeOutput.log".
See also "/root/build/CMakeFiles/CMakeError.log".

~/build # pip3 show protobuf
Name: protobuf
Version: 3.14.0
Summary: Protocol Buffers
Home-page: https://developers.google.com/protocol-buffers/
Author: None
Author-email: None
License: 3-Clause BSD License
Location: /usr/lib/python3.7/site-packages
Requires: six
Required-by:
~/build # pip3 install protobuf
Requirement already satisfied: protobuf in /usr/lib/python3.7/site-packages (3.14.0)
Requirement already satisfied: six>=1.9 in /usr/lib/python3.7/site-packages (from protobuf) (1.15.0)
```
并且我用pip升级protobuf之后，源码还是找不到升级之后的protobuf。

**Pull request closed without merging**